src/Symfony/Bundle/SecurityBundle/Tests/Security/FirewallMapTest.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Tests\Security;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Bundle\SecurityBundle\Security\FirewallContext;
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+
+class FirewallMapTest extends TestCase
+{
+    const ATTRIBUTE_FIREWALL_CONTEXT = '_firewall_context';
+
+    public function testGetListenersWithEmptyMap()
+    {
+        $request = new Request();
+
+        $map = array();
+        $container = $this->getMockBuilder(Container::class)->getMock();
+        $container->expects($this->never())->method('get');
+
+        $firewallMap = new FirewallMap($container, $map);
+
+        $this->assertEquals(array(array(), null), $firewallMap->getListeners($request));
+        $this->assertNull($firewallMap->getFirewallConfig($request));
+        $this->assertFalse($request->attributes->has(self::ATTRIBUTE_FIREWALL_CONTEXT));
+    }
+
+    public function testGetListenersWithInvalidParameter()
+    {
+        $request = new Request();
+        $request->attributes->set(self::ATTRIBUTE_FIREWALL_CONTEXT, 'foo');
+
+        $map = array();
+        $container = $this->getMockBuilder(Container::class)->getMock();
+        $container->expects($this->never())->method('get');
+
+        $firewallMap = new FirewallMap($container, $map);
+
+        $this->assertEquals(array(array(), null), $firewallMap->getListeners($request));
+        $this->assertNull($firewallMap->getFirewallConfig($request));
+        $this->assertFalse($request->attributes->has(self::ATTRIBUTE_FIREWALL_CONTEXT));
+    }
+
+    public function testGetListeners()
+    {
+        $request = new Request();
+
+        $firewallContext = $this->getMockBuilder(FirewallContext::class)->disableOriginalConstructor()->getMock();
+        $firewallContext->expects($this->once())->method('getConfig')->willReturn('CONFIG');
+        $firewallContext->expects($this->once())->method('getContext')->willReturn(array('LISTENERS', 'EXCEPTION LISTENER'));
```

Comment:
While merging into 3.3, [`getContext` is replaced by `getListeners` and `getExceptionListener`](https://github.com/symfony/symfony/blob/3.3/src/Symfony/Bundle/SecurityBundle/Security/FirewallContext.php#L40-L58):
```php
$firewallContext->expects($this->once())->method('getListeners')->willReturn('LISTENERS');
$firewallContext->expects($this->once())->method('getExceptionListener')->willReturn('EXCEPTION LISTENER');
```